File: 13_1-D_Dynamic_Programming/07_Decode_Ways/0091-decode-ways.py
"""
Problem: LeetCode 91 - Decode Ways

Key Idea:
The key idea for solving this problem is to use dynamic programming to count the number of ways to decode the given string. We will create a DP array where dp[i] represents the number of ways to decode the string s[:i] (the first i characters of the string).

Time Complexity:
The time complexity is O(n), where n is the length of the input string s. We iterate through the string once to fill in the DP array.

Space Complexity:
The space complexity is O(n) because we use an additional DP array of the same length as the input string.
"""


class Solution:
    def numDecodings(self, s: str) -> int:
        n = len(s)

        # Initialize a DP array to store the number of ways to decode substrings.
        dp = [0] * (n + 1)

        # Base cases:
        dp[0] = 1  # An empty string can be decoded in one way.
        dp[1] = (
            1 if s[0] != "0" else 0
        )  # The first character can be decoded in one way if it's not '0'.

        # Fill in the DP array.
        for i in range(2, n + 1):
            # Check the one-digit and two-digit possibilities.
            one_digit = int(s[i - 1])
            two_digits = int(s[i - 2 : i])

            # If the one-digit is not '0', it can be decoded in the same way as dp[i-1].
            if one_digit != 0:
                dp[i] += dp[i - 1]

            # If the two-digit is between 10 and 26, it can be decoded in the same way as dp[i-2].
            if 10 <= two_digits <= 26:
                dp[i] += dp[i - 2]

        return dp[n]

File: 13_1-D_Dynamic_Programming/07_Decode_Ways/0091-decode-ways.rs
/*
Problem: LeetCode 91 - Decode Ways

Key Idea:
The key idea is to use dynamic programming to count the number of ways to decode a given string.

Approach:
1. Initialize a vector `dp` of size n+1, where n is the length of the input string.
   - `dp[i]` will represent the number of ways to decode the substring from index i to n-1.
2. Set `dp[n]` to 1, as there is one way to decode an empty string.
3. Iterate through the characters of the input string in reverse order:
   - For each character at index `i`, calculate the number of ways to decode the substring starting from index `i`:
     - If the current character is '0', set `dp[i]` to 0 because '0' cannot be decoded as a single character.
     - Otherwise, set `dp[i]` to `dp[i+1]` because the current character can be decoded as a single character.
     - If the current character and the next character form a valid two-digit number (between '10' and '26'), add `dp[i+2]` to `dp[i]` because the two characters can be decoded as a single unit.
4. Finally, `dp[0]` will contain the number of ways to decode the entire string.
5. Return `dp[0]` as the result.

Time Complexity:
O(n), where n is the length of the input string. We perform a single pass through the string.

Space Complexity:
O(n), as we use a vector of size n+1 to store the dynamic programming values.
*/

impl Solution {
    pub fn num_decodings(s: String) -> i32 {
        let s = s.as_bytes();
        let n = s.len();
        let mut dp = vec![0; n + 1];
        dp[n] = 1;

        for i in (0..n).rev() {
            if s[i] == b'0' {
                dp[i] = 0;
            } else {
                dp[i] = dp[i + 1];
                if i + 1 < n && (s[i] == b'1' || (s[i] == b'2' && s[i + 1] <= b'6')) {
                    dp[i] += dp[i + 2];
                }
            }
        }

        dp[0]
    }
}
